io/jvm/src/main/scala/fs2/io/internal/PipedStreamBuffer.scala

```diff
@@ -364,15 +364,35 @@ private[io] final class PipedStreamBuffer(private[this] val capacity: Int) { sel
         dstCap: Int,
         length: Int
     ): Unit = {
-      val dstOffset = dstPos % dstCap
+      val dstOffset = modulus(dstPos, dstCap)
       if (dstOffset + length >= dstCap) {
         val batch1 = dstCap - dstOffset
         val batch2 = length - batch1
         System.arraycopy(src, srcPos, dst, dstOffset, batch1)
         System.arraycopy(src, srcPos + batch1, dst, 0, batch2)
       } else {
-        System.arraycopy(src, srcPos, dst, dstOffset, length)
+        try System.arraycopy(src, srcPos, dst, dstOffset, length)
+        catch {
+          case e: ArrayIndexOutOfBoundsException =>
+            println(
+              s"srcPos = $srcPos, dstPos = $dstPos, dstCap = $dstCap, length = $length, dstOffset = $dstOffset, dstOffset + length = ${dstOffset + length}"
+            )
+            throw e
+        }
       }
     }
   }
+
+  /** Calculates `n` modulo `m`. This is different from the JVMs built in `%`
+    * remainder operator which can return a negative result for negative
+    * numbers. This method always returns a non-negative result.
+    *
+    * @param n the divident
+    * @param m the divisor
+    * @return the result of `n` modulo `m`
+    */
+  private[this] def modulus(n: Int, m: Int): Int = {
+    val r = n % m
+    if (r < 0) r + m else r
+  }
```

Comment:
Can this be replaced with `lang.Integer.remainderUnsigned`?
https://docs.oracle.com/javase/8/docs/api/java/lang/Integer.html#remainderUnsigned-int-int-

Reply:
That worked. Thanks.